ui/lib/apps/InstanceProfiling/pages/Detail.tsx

```diff
@@ -37,6 +74,119 @@ function isFinished(data) {
   return groupState === 2 || groupState === 3
 }
 
+async function getActionToken(
+  id: string,
+  apiType: string
+): Promise<string | undefined> {
+  const res = await client.getInstance().getActionToken(id, apiType)
+  const token = res.data
+  if (!token) {
+    return
+  }
+  return token
+}
+
+function ViewResultButton({ rec, t }) {
+  const isProtobuf: boolean =
+    rec.profile_raw_data_type === ProfilingRawDataTypeProtobuf
+  let token: string | undefined
+
+  const handleViewResultMenuClick = usePersistFn(async (e: MenuInfo) => {
+    switch (e.key) {
+      case 'download':
+        token = await getActionToken(rec.id, 'single_download')
+        if (!token) {
+          return
+        }
+
+        window.location.href = `${client.getBasePath()}/profiling/single/download?token=${token}`
+        break
+      default:
+        token = await getActionToken(rec.id, 'single_view')
+        if (!token) {
+          return
+        }
+
+        let profileURL = `${client.getBasePath()}/profiling/single/view?token=${token}&output_type=${
+          e.key
+        }`
+        window.open(`${profileURL}`, '_blank')
+    }
+  })
+
+  const handleViewResultBtnClick = usePersistFn(async () => {
+    token = await getActionToken(rec.id, 'single_view')
+    if (!token) {
+      return
+    }
+
+    let profileURL = `${client.getBasePath()}/profiling/single/view?token=${token}`
+    if (isProtobuf) {
+      const titleOnTab = rec.target.kind + '_' + rec.target.display_name
+      profileURL = `/dashboard/speedscope#profileURL=${encodeURIComponent(
+        profileURL + `&output_type=${ViewOutputTypeOptions.FlameGraph}`
+      )}&title=${titleOnTab}`
+    }
+
+    window.open(`${profileURL}`, '_blank')
+  })
+
+  const menu = () => {
+    return (
+      <Menu onClick={handleViewResultMenuClick}>
+        {rec.view_output_type_options.map((option, idx) => {
+          // skip the first option in menu since it has been show on the button.
+          if (idx != 0) {
+            return (
+              <Menu.Item key={option}>
+                {t(
+                  `instance_profiling.detail.table.columns.selection.types.${option}`
+                )}
+              </Menu.Item>
+            )
+          }
+        })}
+      </Menu>
+    )
+  }
+
+  const DropdownButton = () => {
+    return (
+      <Dropdown.Button
+        disabled={rec.state !== taskState.Success}
+        overlay={menu}
+        onClick={handleViewResultBtnClick}
+      >
+        {t(
+          `instance_profiling.detail.table.columns.selection.types.${rec.view_output_type_options[0]}`
+        )}
+      </Dropdown.Button>
+    )
+  }
+
+  return (
+    <>
+      {isProtobuf ? (
+        <DropdownButton />
+      ) : (
```

Comment:
I think your check can be based on `how many view_output_type_options are there`, which is more semantically correct, e.g.

```
if view_output_type_options.length > 1 {
  render as drop down button
} else {
  render as a single button 
}
```

Since the core reason that you want to display a drop down button is that there are multiple view outputs.

This also avoids you checking the underlying raw data type multiple places. You have already checked once in the `mapData()`, which is enough. The more places you introduce these logics, the more complex your code becomes.